kaspresso/src/main/kotlin/com/kaspersky/kaspresso/interceptors/behavior/impl/autoscroll/FallbackAutoScrollToAction.kt

```diff
@@ -0,0 +1,91 @@
+package com.kaspersky.kaspresso.interceptors.behavior.impl.autoscroll
+
+import android.util.Log
+import android.view.View
+import android.view.ViewParent
+import android.widget.HorizontalScrollView
+import android.widget.ListView
+import android.widget.ScrollView
+import androidx.core.widget.NestedScrollView
+import androidx.test.espresso.PerformException
+import androidx.test.espresso.UiController
+import androidx.test.espresso.ViewAction
+import androidx.test.espresso.matcher.ViewMatchers
+import androidx.test.espresso.util.HumanReadables
+import org.hamcrest.Matcher
+import org.hamcrest.Matchers
+
+class FallbackAutoScrollToAction : ViewAction {
+    override fun getConstraints(): Matcher<View> {
+        return Matchers.allOf(
+            ViewMatchers.withEffectiveVisibility(ViewMatchers.Visibility.VISIBLE),
+            ViewMatchers.isDescendantOfA(
+                Matchers.anyOf(
+                    ViewMatchers.isAssignableFrom(NestedScrollView::class.java),
+                    ViewMatchers.isAssignableFrom(ScrollView::class.java),
+                    ViewMatchers.isAssignableFrom(HorizontalScrollView::class.java),
+                    ViewMatchers.isAssignableFrom(ListView::class.java)
+                )
+            )
+        )
+    }
+
+    private fun View?.isScrollable(): Boolean =
+        (this is ScrollView || this is NestedScrollView || this is HorizontalScrollView || this is ListView)
+
+    private tailrec fun View?.findFirstParentScrollableView(lastView: View): View? {
+        return if (this == lastView) {
+            if (this.isScrollable()) this else null
+        } else {
+            if (this.isScrollable()){
+                this
+            } else {
+                this?.parent?.findFirstView()?.findFirstParentScrollableView(lastView)
+            }
+        }
+    }
+
+    private tailrec fun ViewParent?.findFirstView(): View? {
+        return if (this is View) {
+            this
+        } else {
+            this?.parent.findFirstView()
+        }
+    }
+
+    override fun perform(uiController: UiController, view: View) {
+        if (ViewMatchers.isDisplayingAtLeast(100).matches(view)) {
+            Log.i(TAG, "View is already displayed. Returning.")
+            return
+        }
+
+        val scrollView = view.findFirstParentScrollableView(view.rootView)
+
+        when (scrollView is HorizontalScrollView) {
+            true -> scrollView.scrollTo(view.right + scrollView.paddingEnd, 0)
+            false -> scrollView?.scrollTo(0, view.top - scrollView.paddingTop)
+        }
+
+        uiController.loopMainThreadUntilIdle()
+        if (!ViewMatchers.isDisplayingAtLeast(100).matches(view)) {
```

Comment:
What does value "100" mean?

Reply:
That the view is 100% visible

Reply:
I mean it's better to use a constant with an relative name